### ISSUE TITLE
[Bug] vllm vs sglang performance test comparison

### ISSUE THREAD

**Issue by luhairong11** [sgl-project/sglang#4245]
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

### Describe the bug

Why is the average time to first token for VLLM significantly faster than that of SGLang, by nearly 10 times? Could it be that the command parameters I used to start SGLang are incorrect?

## test script
```python
evalscope perf \
    --url "http://ip:31008/v1/chat/completions" \
    --parallel 100 \
    --model qwen2.5-32b \
    --number 500\
    --api openai \
    --dataset openqa \
    --stream
```

## vllm command 
```python
CUDA_VISIBLE_DEVICES=4 vllm serve /data/chdmx/models/LLM_models/qwen/Qwen2.5/Qwen2.5-32B-Instruct --tensor-parallel-size 1 --gpu-memory-utilization 0.9 --max-model-len 20000 --trust-remote-code --host 0.0.0.0 --port 31008 --served-model-name qwen2.5-32b
```
**result:**

![Image](https://github.com/user-attachments/assets/25c794d6-bfdd-4e8c-bb8c-6a3e5e621d45)

## sglang command 
```python
CUDA_VISIBLE_DEVICES=4 python3 -m sglang.launch_server --model-path /data/chdmx/models/LLM_models/qwen/Qwen2.5/Qwen2.5-32B-Instruct --tp 1 --mem-fraction-static 0.9 --context-length 20000 --trust-remote-code --host 0.0.0.0 --port 31008
```
**result:**

![Image](https://github.com/user-attachments/assets/5583fcb4-8f77-4fec-ad68-7a6d5ad45be2)

### Environment

root@bms-schyjdmx03:/sgl-workspace# python3 -m sglang.check_env
2025-03-09 19:03:57,307 - INFO - flashinfer.jit: Prebuilt kernels not found, using JIT backend
INFO 03-09 19:04:03 __init__.py:190] Automatically detected platform cuda.
/usr/local/lib/python3.10/dist-packages/torch/utils/cpp_extension.py:1964: UserWarning: TORCH_CUDA_ARCH_LIST is not set, all archs for visible cards are included for compilation. 
If this is not desired, please set os.environ['TORCH_CUDA_ARCH_LIST'].
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/pydantic/_internal/_config.py:345: UserWarning: Valid config keys have changed in V2:
* 'fields' has been removed
  warnings.warn(message, UserWarning)
Python: 3.10.16 (main, Dec  4 2024, 08:53:37) [GCC 9.4.0]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA A100 80GB PCIe
GPU 0,1,2,3,4,5,6,7 Compute Capability: 8.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.4, V12.4.131
CUDA Driver Version: 550.54.15
PyTorch: 2.5.1+cu124
sgl_kernel: 0.0.4
flashinfer: 0.2.2.post1
triton: 3.1.0
transformers: 4.48.3
torchao: 0.8.0
numpy: 1.26.4
aiohttp: 3.11.11
fastapi: 0.115.6
hf_transfer: 0.1.9
huggingface_hub: 0.29.1
interegular: 0.3.3
modelscope: 1.22.3
orjson: 3.10.15
packaging: 24.2
psutil: 6.1.1
pydantic: 2.10.5
multipart: 0.0.12
zmq: 26.2.0
uvicorn: 0.34.0
uvloop: 0.21.0
vllm: 0.7.2
openai: 1.59.8
anthropic: 0.43.1
decord: 0.6.0
NVIDIA Topology: 
        GPU0    GPU1    GPU2    GPU3    GPU4    GPU5    GPU6    GPU7    NIC0    NIC1    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      PIX     PIX     PIX     SYS     SYS     SYS     SYS     SYS     SYS     0-27,56-83      0               N/A
GPU1    PIX      X      PIX     PIX     SYS     SYS     SYS     SYS     SYS     SYS     0-27,56-83      0               N/A
GPU2    PIX     PIX      X      PIX     SYS     SYS     SYS     SYS     SYS     SYS     0-27,56-83      0               N/A
GPU3    PIX     PIX     PIX      X      SYS     SYS     SYS     SYS     SYS     SYS     0-27,56-83      0               N/A
GPU4    SYS     SYS     SYS     SYS      X      PIX     PIX     PIX     SYS     SYS     28-55,84-111    1               N/A
GPU5    SYS     SYS     SYS     SYS     PIX      X      PIX     PIX     SYS     SYS     28-55,84-111    1               N/A
GPU6    SYS     SYS     SYS     SYS     PIX     PIX      X      PIX     SYS     SYS     28-55,84-111    1               N/A
GPU7    SYS     SYS     SYS     SYS     PIX     PIX     PIX      X      SYS     SYS     28-55,84-111    1               N/A
NIC0    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS      X      SYS
NIC1    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS      X 

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_0
  NIC1: mlx5_1

ulimit soft: 1048576

**Comment by Swipe4057** [comment]
For sglang, I always add these parameters: --enable-mixed-chunk --enable-torch-compile

And for vllm, I use these parameters: --enable-prefix-caching --enable-chunked-prefill

Moreover, the parameters gpu-memory-utilization and mem-fraction-static are not equivalent to each other, they allocate memory differently, so the comparison is not accurate. Please check the documentation for details.

**Comment by luhairong11** [comment]
## For sglang, I add parameters: --enable-mixed-chunk --enable-torch-compile
**result:**

![Image](https://github.com/user-attachments/assets/d23af718-a83b-4f6b-9d19-9a86308b9162)

## for vllm, I add parameters: --enable-prefix-caching --enable-chunked-prefill
**result:**

![Image](https://github.com/user-attachments/assets/4dfab4ff-0ea4-47d6-94b3-e4906f2c856b)

From the results above, it can be seen that there is still a significant difference in the average time to first token.

**Comment by Swipe4057** [comment]
There may be other differences in default settings:
https://docs.vllm.ai/en/latest/models/performance.html
https://docs.sglang.ai/backend/hyperparametertuning.html

For example, max-num-batched-tokens in vllm is 2048 by default, while in sglang it's called --chunked-prefill-size and is set to 8192. Also, the default scheduler in sglang is lpm while in vllm it's fcfs. Default settings are used if you don't specify them explicitly, and there may be other differences that I haven't mentioned. In any case, this might not be a bug, but rather a characteristic of the current state of the frameworks, as they are both evolving and not standing still, so different performance is not a bug. Moreover, according to your statistics, the overall throughput of sglang is still higher in both the first and second cases. I think more experiments should be conducted.
But I'm just a regular user

**Comment by luhairong11** [comment]
Yes, this is not a bug. I simply want to assess which of the two frameworks, vllm and sglang, is faster during inference for potential use in engineering applications. The results above indicate that vllm still has a significant advantage in terms of average time to first token.

**Comment by Swipe4057** [comment]
If you're only concerned about time to first token, try changing the sglang scheduling policy from lpm (which is oriented toward maximum throughput) to fcfs (first come, first served).

**Comment by qiulang** [comment]
@Swipe4057 @luhairong11  @minleminzui ，

I did some similar benchmark test myself and find something hard to understand, for single-GPU configuration, SGLang demonstrated several impressive performance characteristics for small LLM (Qwen 2.5 7B AWQ). But when I do the same test in 2 GPUs configuration and 4 GPUs configuration, SGLang performances surprisingly degraded significantly  rather than improved.

I can't figure out why so I opened in issue  https://github.com/sgl-project/sglang/issues/5808 

Maybe you guys can help solve it.

**Comment by mlleo** [comment]
Can you let me know the version of vllm and sglang that you tested?
